sphinx/addnodes.py

```diff
@@ -144,18 +172,63 @@ def child_text_separator(self):
 
 
 class desc_signature_line(nodes.Part, nodes.Inline, nodes.FixedTextElement):
-    """Node for a line in a multi-line object signatures.
+    """Node for a line in a multi-line object signature.
 
-    It should only be used in a ``desc_signature`` with ``is_multiline`` set.
+    It should only be used as a child of a :py:class:`desc_signature`
+    with ``is_multiline`` set to ``True``.
     Set ``add_permalink = True`` for the line that should get the permalink.
     """
     sphinx_line_type = ''
 
 
+class desc_content(nodes.General, nodes.Element):
+    """Node for object description content.
+
+    Must be the last child node in a :py:class:`desc` node.
+    """
+
+
+class desc_inline(_desc_classes_injector, nodes.Inline, nodes.TextElement):
+    """Node for a signature fragment in inline text.
+
+    This is for example used for roles like :rst:role:`cpp:expr`.
+
+    This node always has the classes ``sig``, ``sig-inline``,
+    and the name of the domain it belongs to.
+    """
+    classes = ['sig', 'sig-inline']
+
+    def __init__(self, domain: str, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self['classes'].append(domain)
+
+
+# Nodes for high-level structure in signatures
+##############################################
+
 # nodes to use within a desc_signature or desc_signature_line
 
-class desc_addname(nodes.Part, nodes.Inline, nodes.FixedTextElement):
-    """Node for additional name parts (module name, class name)."""
+class desc_name(_desc_classes_injector, nodes.Part, nodes.Inline, nodes.FixedTextElement):
+    """Node for the main object name.
+
+    For example, in the declaration of a Python class ``MyModule.MyClass``,
+    the main name is ``MyClass``.
+
+    This node always has the class ``sig-name``.
+    """
+    classes = ['sig-name', 'descname']  # 'descname' is for backwards compatibility
+
+
+class desc_addname(_desc_classes_injector, nodes.Part, nodes.Inline, nodes.FixedTextElement):
```

Comment:
Adding a new node brings an incompatible change for 3rd party writers. It would be better to provide a fallback to them like `SigElementFallbackTransform` to keep them working (for a while).

Reply:
This is fortunately not a new node, I just moved them a bit around so the order better fits the purpose of them.

Reply:
It seems this comment has been drifted. Originally, I commented to the `desc_inline` node.

Reply:
Ah, makes sense. That addition to ``SigElementFallbackTransform`` should already be there. I change its logic such that:
- if there is no ``desc_inline`` handler, then transform them.
- (as before) if not all ``desc_sig_element`` have handlers, then transform them all.

Reply:
Reasonable. +1 for the change.